### PR TITLE
HPCC-28034 Prioritize lingering Thor over agent

### PR DESCRIPTION
**Pull request by jakesmith**
## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: ecl/agentexec/agentexec.cpp [hunk 1]
```diff
@@ -212,20 +221,25 @@ int CEclAgentExecutionServer::run()
 
 //---------------------------------------------------------------------------------
 
-typedef std::tuple<unsigned, const char *, const char *> ThreadCtx;
+typedef std::tuple<IJobQueueItem *, unsigned, const char *, const char *> ThreadCtx;
 
 // NB: WaitThread only used by if pool created (see CEclAgentExecutionServer ctor)
 class WaitThread : public CInterfaceOf<IPooledThread>
 {
+    CEclAgentExecutionServer &owner;
```
Comment (ghalliday) [inline ecl/agentexec/agentexec.cpp:229]: Confusing to split the member variables - these should go with the others at the end.

File: common/workunit/wujobq.cpp [hunk 2]
```diff
@@ -2152,6 +2152,47 @@ extern bool WORKUNIT_API runWorkUnit(const char *wuid, const char *queueName)
...
+            // if now not there to be removed, implies dequeue
+            if (!queue->remove(job))
+                consumed = true;
+        }
+        if (!consumed)
+        {
+            // thought some were waiting, but still on queue, validate queue client sessions
+            queue->connect(true);
+            unsigned connected, waiting, enqueued;
+            queue->getStats(connected, waiting, enqueued);
```
Comment (ghalliday) [inline common/workunit/wujobq.cpp:2189]: Is this needed if true has been passed to connect()?

File: ecl/agentexec/agentexec.cpp [hunk 3]
```diff
@@ -253,25 +267,28 @@ class WaitThread : public CInterfaceOf<IPooledThread>
             bool useChildProcesses = compConfig->getPropBool("@useChildProcesses");
             if (isContainerized() && !useChildProcesses)
             {
-                std::list<std::pair<std::string, std::string>> params = { };
-                if (compConfig->getPropBool("@useThorQueue", true))
-                    params.push_back({ "queue", queue.get() });
-                StringBuffer jobName(wuid);
-                if (isThorAgent)
+                if (!owner.lingerQueue || !queueJobIfQueueWaiting(owner.lingerQueue, item, 10000, 1000))
```
Comment (ghalliday) [inline ecl/agentexec/agentexec.cpp:270]: Worth exposing the two timeouts as named constants to improve readability.  There is a possibility they might be configured in the future.

File: ecl/agentexec/agentexec.cpp [hunk 4]
```diff
@@ -212,20 +221,25 @@ int CEclAgentExecutionServer::run()
...
+        : owner(_owner), dali(_dali), apptype(_apptype), queue(_queue)
     {
         isThorAgent = streq("thor", apptype);
     }
     virtual void init(void *param) override
     {
         auto &context = *static_cast<ThreadCtx *>(param);
-        std::tie(wfid, wuid, graphName) = context;
+        IJobQueueItem *_item;
+        std::tie(_item, wfid, wuid, graphName) = context;
```
Comment (ghalliday) [inline ecl/agentexec/agentexec.cpp:241]: If item was a `Linked<IJobQueueItem>` could `item` be used in the tie as well (avoiding the local variable)?

**Review by ghalliday — COMMENTED** [review]
@jakesmith Looks good.  A few minor comments/questions.

File: common/workunit/wujobq.cpp (hunk 2 above)
Comment (jakesmith) [reply common/workunit/wujobq.cpp:2189]: ah, no you're right, connect does it anyway if validateitemsessions = true
will remove.

File: ecl/agentexec/agentexec.cpp (hunk 1 above)
Comment (jakesmith) [reply ecl/agentexec/agentexec.cpp:229]: will move them.

File: ecl/agentexec/agentexec.cpp (hunk 4 above)
Comment (jakesmith) [reply ecl/agentexec/agentexec.cpp:241]: not solely.. the tuple is assigning by reference, for it to work the IJobQueueItem in the tuple needs to be a Linked as well

Not sure it's better, but I've changed it.

**Comment by jakesmith** [comment]
@ghalliday - please review new commit

File: ecl/agentexec/agentexec.cpp (hunk 4 above)
Comment (ghalliday) [reply ecl/agentexec/agentexec.cpp]: I have never used tie, so I didn't know that.  wuid is a const char * in the tuple, but StringAttr in the class, so I didn't think it needed to be the same type, only assignable.

**Review by ghalliday — APPROVED** [review]
Please squash

**Comment by jakesmith** [comment]
@ghalliday - squashed

**Pull request merged**